### PR TITLE
storage: support to get old value through DeltaScanner

### PR DESCRIPTION
**Pull request by 5kbpers**
Signed-off-by: 5kbpers <tangminghua@pingcap.com>

### What problem does this PR solve?

Problem Summary:
Old Value is the previously written value before a delete/update operation.
To support the new collation format and compatible with TiDB Binlog, TiCDC requires TiKV to output the old value for it.

### What is changed and how it works?
Separate from https://github.com/tikv/tikv/pull/8071

What's Changed:
* Add a flag to `DeltaScanner`
* Look-ahead for seeking old value in `DeltaScanner`

### Related changes

- Should be cherry-picked to release 4.0 after all related PRs got merged.

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM

### Release note 
* N/A

### CODE REVIEW

File: src/storage/mvcc/reader/scanner/forward.rs [hunk 1]
```diff
@@ -558,14 +560,59 @@ fn scan_latest_handle_lock<S: Snapshot, T>(
...
+                        default_cursor,
+                        user_key,
+                        write_ref.start_ts,
+                        statistics,
+                    )?)
+                };
+                break;
+            }
+            WriteType::Delete => break,
+            WriteType::Lock | WriteType::Rollback => {
```
Comment (sticnarf) [inline src/storage/mvcc/reader/scanner/forward.rs:607]: It seems in this PR we will skip some `Lock` for reading the old value. However, the current implementation outputs `Lock`. Will this behavior change affect CDC? @MyonKeminta

File: components/cdc/src/endpoint.rs [hunk 2]
```diff
@@ -685,7 +686,7 @@ impl Initializer {
         let current = TimeStamp::max();
         let mut scanner = ScannerBuilder::new(snap, current, false)
             .range(None, None)
-            .build_delta_scanner(self.checkpoint_ts)
+            .build_delta_scanner(self.checkpoint_ts, ExtraOp::ReadOldValue)
```
Comment (overvenus) [inline components/cdc/src/endpoint.rs:689]: Looks like it reads old value by default, is it expected?

File: src/storage/mvcc/reader/scanner/forward.rs [hunk 3]
```diff
@@ -598,10 +601,26 @@ impl<S: Snapshot> ScanPolicy<S> for DeltaEntryPolicy {
             } else {
                 Ok((vec![], vec![]))
             };
+            let old_value = if self.extra_op == ExtraOp::ReadOldValue
```
Comment (overvenus) [inline src/storage/mvcc/reader/scanner/forward.rs:604]: Does it affect incremental backup?

File: src/storage/mvcc/reader/scanner/mod.rs [hunk 4]
```diff
@@ -146,7 +146,11 @@ impl<S: Snapshot> ScannerBuilder<S> {
         ))
     }
 
-    pub fn build_delta_scanner(mut self, from_ts: TimeStamp) -> Result<DeltaScanner<S>> {
+    pub fn build_delta_scanner(
+        mut self,
+        from_ts: TimeStamp,
+        extra_op: ExtraOp,
+    ) -> Result<DeltaScanner<S>> {
```
Comment (overvenus) [inline src/storage/mvcc/reader/scanner/mod.rs:153]: How about adding a new method `fn extra_op()`?

File: components/cdc/src/endpoint.rs (hunk 2 above)
Comment (5kbpers) [reply components/cdc/src/endpoint.rs:689]: My fault.😢

File: src/storage/mvcc/reader/scanner/forward.rs (hunk 3 above)
Comment (5kbpers) [reply src/storage/mvcc/reader/scanner/forward.rs:604]: We should set `self.extra_op` to `ExtraOp::Noop` in incremental backup.

File: src/storage/mvcc/reader/scanner/mod.rs (hunk 4 above)
Comment (5kbpers) [reply src/storage/mvcc/reader/scanner/mod.rs:153]: We only need this flag for `DeltaEntryPolicy`.

File: src/storage/mvcc/reader/scanner/mod.rs [hunk 5]
```diff
@@ -359,6 +363,57 @@ pub fn has_data_in_range<S: Snapshot>(
...
+    while write_cursor.valid()?
+        && Key::is_user_key_eq(
+            write_cursor.key(&mut statistics.write),
+            user_key.as_encoded(),
+        )
+    {
+        assert_ge!(
+            after_ts,
+            Key::decode_ts_from(write_cursor.key(&mut statistics.write))?
+        );
```
[the remark this answers is not in the document]
Comment (5kbpers) [reply src/storage/mvcc/reader/scanner/mod.rs:388]: `Lock::Pessimistic` have been skipped here. see forward.rs#L605

[the remark this answers is not in the document]
Comment (MyonKeminta) [reply src/storage/mvcc/reader/scanner/mod.rs:388]: A pessimistic transaction's pessimistic lock will first be replaced by non-pessimistic locks, and then be committed, and its start_ts won't change.

[the remark this answers is not in the document]
Comment (MyonKeminta) [reply src/storage/mvcc/reader/scanner/mod.rs:388]: You can add a case like this to unit tests:
* Prewrite start_ts = 1
* Commit start_ts = 10
* Acquire pessimistic lock start_ts = 5, for_update_ts = 11
* Pessimistic prewrite start_ts = 5, for_update_ts = 11
* Scan

**Comment by sticnarf** [comment]
LGTM cancel

**Comment by sticnarf** [comment]
LGTM

**Comment by AndreMouche** [comment]
LGTM

**Comment by AndreMouche** [comment]
LGTM cancel

**Comment by AndreMouche** [comment]
lgtm cancel

**Comment by AndreMouche** [comment]
lgtm cancel

**Comment by AndreMouche** [comment]
LGTM

**Comment by AndreMouche** [comment]
LGTM cancel

**Comment by AndreMouche** [comment]
lgtm cancel

**Comment by AndreMouche** [comment]
lgtm cancel

**Comment by AndreMouche** [comment]
lgtm cancel

**Comment by AndreMouche** [comment]
LGTM

**Comment by AndreMouche** [comment]
LGTM

**Comment by AndreMouche** [comment]
lgtm cancel

**Comment by AndreMouche** [comment]
LGTM

**Comment by AndreMouche** [comment]
lgtm cancel

**Comment by AndreMouche** [comment]
LGTM

**Comment by AndreMouche** [comment]
lgtm

**Comment by AndreMouche** [comment]
lgtm

**Comment by AndreMouche** [comment]
LGTM

**Comment by AndreMouche** [comment]
LGTM

**Comment by AndreMouche** [comment]
lgtm cancel

**Comment by AndreMouche** [comment]
lgtm cancel

**Comment by AndreMouche** [comment]
LGTM

**Comment by AndreMouche** [comment]
lgtm

**Comment by AndreMouche** [comment]
lgtm cancel

**Comment by AndreMouche** [comment]
LGTM

**Pull request merged**